apps/backend/user/src/login.ts

```diff
@@ -0,0 +1,57 @@
+import { Request, Response } from "express";
+import { IDatabase } from "pg-promise";
+import { User } from "./models/user";
+import bcrypt from "bcryptjs";
+
+const login = async (req: Request, res: Response, db: IDatabase<object>) => {
+  const { email, password } = req.body;
+
+  if (!email || !password) {
+    return res.status(400).json({ error: "Email and password are required" });
+  }
+
+  try {
+    const user = await db.oneOrNone<User>(
+      "SELECT * FROM users WHERE email = $1",
+      [email],
+    );
+
+    if (!user) {
+      return res.status(401).json({ error: "Invalid email or password" });
+    }
```

Comment:
This exposes us to a timing attack. An attacker can try email/password combos and can check whether or not an email exists in the db by how long the response takes (short if not signed up since it returns early, long if the email exists since then more cryptographic operations are performed on the password).

The easiest way to mitigate this is to always make a bcrypt comparison either against the hashed password in the db or a static string of a random value.

Reply:
I considered this and the old issue where the error messages distinguished between wrong username and wrong password. I don't think it matters if people know a username/email is in the DB